mrmustard/lab_dev/samplers.py

```diff
@@ -217,24 +217,30 @@ def probabilities(self, state, atol=1e-4):
         )
         return self._validate_probs(probs, atol)
 
-    def sample(self, state: State, n_samples: int = 1000, seed: int | None = None) -> np.ndarray:
-        initial_mode = state.modes[0]
-        initial_samples, probs = self.sample_prob_dist(state[initial_mode], n_samples, seed)
+    def sample_prob_dist(
+        self, state: State, n_samples: int = 1000, seed: int | None = None
+    ) -> np.ndarray:
+        r"""
+        Samples a state by computing the probability distribution.
 
-        if len(state.modes) == 1:
-            return initial_samples
+        Args:
+            state: The state to sample.
+            n_samples: The number of samples to generate.
+            seed: An optional seed for random sampling.
 
-        unique_samples, counts = np.unique(initial_samples, return_counts=True)
-        ret = []
-        for unique_sample, counts in zip(unique_samples, counts):
-            quad = np.array([[unique_sample] + [None] * (state.n_modes - 1)])
-            quad = quad if isinstance(state, Ket) else math.tile(quad, (1, 2))
-            reduced_rep = (state >> BtoQ([initial_mode], phi=self._phi)).representation(quad)
-            reduced_state = state.__class__.from_bargmann(state.modes[1:], reduced_rep.triple)
-            prob = probs[initial_samples.tolist().index(unique_sample)] / self._step
-            norm = math.sqrt(prob) if isinstance(state, Ket) else prob
-            normalized_reduced_state = reduced_state / norm
-            samples = self.sample(normalized_reduced_state, counts)
-            for sample in samples:
-                ret.append(np.append([unique_sample], sample))
-        return np.array(ret)
+        Returns:
+            A tuple of the generated samples and the probability
+            of obtaining the sample.
+        """
+        rng = np.random.default_rng(seed) if seed else settings.rng
+        probs = self.probabilities(state)
+        meas_outcomes = list(product(self.meas_outcomes, repeat=len(state.modes)))
+        samples = rng.choice(
+            a=meas_outcomes,
+            p=probs,
+            size=n_samples,
+        )
+        return samples
+
+    def sample(self, state: State, n_samples: int = 1000, seed: int | None = None) -> np.ndarray:
+        return self.sample_prob_dist(state, n_samples, seed)
```

Comment:
Are you doing this just to speed up tests or is this more efficient in general?

Reply:
No, I think for CV measurements it's better, because np.unique on a bunch of floats will always give you back the same list, it's nearly impossible that more than one are the same. So it was inefficient to keep calling sample and get a single sample back

Reply:
Hmm is that true in general? Even printing out `counts` on develop I can see that it's not just a list of `1`s. I guess it would depend both on the discretization and the state you're computing with?

I wonder if we should profile the two methods. I did some light profiling when developing and I seem to recall the shortcut did work really well on high n_modes but was comparable to `sample_prob_dist` on lower numbers

Reply:
In the np backend it's fast, but it was painfully slow in the TF backend. This is slightly faster in the np backend too. But yes, we can think of an even better way. This for now is ~50x faster in the TF backend and ~2x faster for np.